websock/session.nim

```diff
@@ -116,13 +119,86 @@ proc send*(
     else:
       default(MaskKey)
 
-  if opcode in {Opcode.Text, Opcode.Cont, Opcode.Binary}:
-    await ws.writeMessage(
-      data, opcode, maskKey, ws.extensions)
+  if opcode.isControl:
+    await ws.writeControl(data, opcode, maskKey)
+  else:
+    await ws.writeMessage(data, opcode, maskKey, ws.extensions)
 
-    return
+proc doSend(
+  ws: WSSession,
+  data: seq[byte] = @[],
+  opcode: Opcode,
+  fut: Future[void]): Future[void]
+  {.raises: [Defect, WSClosedError].} =
+  if fut.cancelled:
+    let sendFut = newFuture[void]("doSend")
+    sendFut.cancel()
+    return sendFut
+
+  let sendFut = ws.nonCancellableSend(data, opcode)
+
+  proc handleSent(future: pointer) =
+    if fut.finished:
+      return
+    if sendFut.failed:
+      fut.fail(sendFut.error)
+    elif sendFut.cancelled:
+      fut.cancel()
+    else:
+      fut.complete()
+
+  sendFut.addCallback(handleSent)
```

Comment:
This style of async (callback) is generally discouraged, especially in this case where you can just
```nim
await sendFut # maybe wrap in try

# move handleSent content here
```

Reply:
The goal here is to unlink `fut` from `sendfut`, so that `sendFut` does not get cancelled when caller cancels `fut`.

With `await` syntax it would not really get clearer, as the result still needs to be propagated from `sendFut` to `fut`, in any case. There's also the early exit `if fut.cancelled`, which is tricky to implement from `async` function (which this is not), as `newCancelledError` is not exported.

Reply:
Indeed currently it's a bit of a pain to do this, though it will improve with a `awaitnc` or similar
Anyway, here is a branch with saner async usage, imo: https://github.com/status-im/nim-websock/commit/a4263867cc42903d64eaa7ccbba343cac22af38c

Reply:
OK, I'm not sure how the performance characteristics change with `try` / `except` here (seems reasonable as a 1 per message overhead), and it's a bit less defensive in a way that it translates `CancelledError` into `fail()` instead of `cancel()` and does not deal with client messing with returned fut in other ways than running `cancel()` on it, but generally it still looks correct to me when used correctly. Will give it a shot as part of https://github.com/status-im/nimbus-eth2/pull/4061

Reply:
> it's a bit less defensive in a way that it translates CancelledError into fail() instead of cancel()

Unless I'm missing something, these can't be cancelled anyway, so CancelledError shouldn't happen?

> does not deal with client messing with returned fut in other ways than running `cancel()` on it

imo the only two possible operations are `await` it or `cancel` it, is there another operation you would expect?

Reply:
When normally used by a caller, that's correct. But it's a library, so there's nothing preventing API misuse. There's also this little extra mental step when reviewing to see "oh _that's_ why it's fine to not propagate cancellation". It's fine by me either way, under normal operation it should work equally well.

Reply:
Right, only `send` is exported, so it shouldn't be possible to misuse as a lib, but ofc we need to be careful to keep it sensible here. Hopefully we'll have `awaitnc` sooner than later :)

Reply:
The `retFut` is also exported, client could `fail(customError)` on it, then we crash because we try to complete it once more. And the `sendFut` is exported via types module. But agree, it's alright for now.

Reply:
Good points!